src/Assistant.ts

```diff
@@ -0,0 +1,376 @@
+T & AllMiddlewareArgs;
+
+/** Constants */
+const ASSISTANT_PAYLOAD_TYPES = new Set(['assistant_thread_started', 'assistant_thread_context_changed', 'message']);
+
+export class Assistant {
+  private threadContextStore: AssistantThreadContextStore;
+
+  /** 'assistant_thread_started' */
+  private threadStarted: AssistantThreadStartedMiddleware[];
+
+  /** 'assistant_thread_context_changed' */
+  private threadContextChanged: AssistantThreadContextChangedMiddleware[];
+
+  /** 'message' */
+  private userMessage: AssistantUserMessageMiddleware[];
+
+  public constructor(config: AssistantConfig) {
+    validate(config);
+
+    const {
+      threadContextStore = new DefaultThreadContextStore(),
+      threadStarted,
+      // When `threadContextChanged` method is not provided, fallback to
+      // AssistantContextStore's save method. If a custom store has also not
+      // been provided, the default save context-via-metadata approach is used.
+      // See DefaultThreadContextStore for details of this implementation.
+      threadContextChanged = (args) => threadContextStore.save(args),
+      userMessage,
+    } = config;
+
+    this.threadContextStore = threadContextStore;
+    this.threadStarted = Array.isArray(threadStarted) ? threadStarted : [threadStarted];
+    this.threadContextChanged = Array.isArray(threadContextChanged) ? threadContextChanged : [threadContextChanged];
+    this.userMessage = Array.isArray(userMessage) ? userMessage : [userMessage];
+  }
+
+  public getMiddleware(): Middleware<AnyMiddlewareArgs> {
+    return async (args): Promise<void> => {
+      if (isAssistantEvent(args) && matchesConstraints(args)) return this.processEvent(args);
+      return args.next();
+    };
+  }
+
+  private async processEvent(args: AllAssistantMiddlewareArgs): Promise<void> {
+    const { payload } = args;
+    const assistantArgs = prepareAssistantArgs(this.threadContextStore, args);
+    const assistantMiddleware = this.getAssistantMiddleware(payload);
+    return processAssistantMiddleware(assistantArgs, assistantMiddleware);
+  }
+
+  /**
+   * `getAssistantMiddleware()` returns the Assistant instance's middleware
+   */
+  private getAssistantMiddleware(payload: AllAssistantMiddlewareArgs['payload']): AssistantMiddleware {
+    switch (payload.type) {
+      case 'assistant_thread_started':
+        return this.threadStarted;
+      case 'assistant_thread_context_changed':
+        return this.threadContextChanged;
+      case 'message':
+        return this.userMessage;
+      default:
+        return [];
+    }
+  }
+}
+
+/**
+ * `prepareAssistantArgs()` takes the event arguments and:
+ *  1. Removes the next() passed in from App-level middleware processing, thus preventing
+ *  events from continuing down the global middleware chain to subsequent listeners
+ *  2. Adds assistant-specific utilities (i.e., helper methods)
+ * */
+export function prepareAssistantArgs(
```

Comment:
May I ask we rename this from `prepare` to `enrich`? Anything named "prepare" is a pet peeve of mine as the definition for what "preparation" means requires knowing the context. We name other such middleware-argument-enriching processes elsewhere ([like in the deno-sdk](https://github.com/slackapi/deno-slack-sdk/blob/834464dd93f69e546567862e4c6d8f1364f93477/src/functions/enrich-context.ts#L7)) enriching. 🙏 

Also, AFAICT this function only gets used once, so let's not `export` it - less API surface area to possibly break in the future. Could even just move it to where it is used (inside `processEvent` here).

Reply:
I export these only for testing purposes! Sounds like that's an outdated approach? If it's possible to test these without exporting, that'll allow me to get rid of the export.

And sue, a very reluctant OK on the rename. 😛

Reply:
oic, no tests yet so you are not importing them yet, is that it?

Reply:
Nope, `prepareAssistantArgs` is definitely used in tests - it has it's [own block](https://github.com/slackapi/bolt-js/pull/2286/files#diff-b31a4d458126a21e6d1d7d7700b3a6802b3b6282cac91ab1973223bdede5e7d1R227)! I imagine the others are there, too. They're destructured from a [parent import](https://github.com/slackapi/bolt-js/pull/2286/files#diff-b31a4d458126a21e6d1d7d7700b3a6802b3b6282cac91ab1973223bdede5e7d1R21).

Reply:
Ah thanks for pointing that out. OK, nevermind 😞  The async imports must have confused my IDE search functionality.

Reply:
No worries. I'm doing a pass just to make sure we're not exporting anything unnecessarily for good measure.